proposals/lip-0053.md

```diff
@@ -252,178 +479,192 @@ crossChainUpdateTransactionParams = {
-
-#### InboxUpdate Validity
-
-If `params` contains a non-empty `certificate` and an `inboxUpdate`, the validity of the latter is checked by:
-
-* Let `newInboxRoot`, `newInboxAppendPath` and `newInboxSize` be the resulting new root, append path and size of the Merkle tree that would be obtained if all messages from `crossChainMessages` would be appended to the `partnerChannel` inbox. Note that `partnerChannel.inbox` is not updated here.
-* If `inboxUpdate` contains a non-empty `messageWitness`, then update `newInboxRoot` to the output of `calculateRootFromRightWitness(newInboxSize, newInboxAppendPath, inboxUpdate.messageWitness.siblingHashes)` as specified in [LIP 0031](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md).
-* Then validate the newly updated root against the certificate state root using the provided `outboxRootWitness`. Using notation from [LIP 0039][lip-0039], this is done via the function `verify(queryKeys, proof, certificate.stateRoot)` with:
-
-```python
-queryKeys = [outboxKey],
-
-proof = {
-    siblingHashes: outboxRootWitness.siblingHashes,
-    queries: [{
-        key: outboxKey,
-        value: newInboxRoot,
-        bitmap: outboxRootWitness.bitmap
-    }],
-}
-
-outboxKey = MODULE_ID_INTEROPERABILITY
-            || STORE_PREFIX_OUTBOX
-            || SHA-256(uint32be(partnerChainID)).
+def execute(trs: Transaction) -> None:
+    # Verify certificate signature. We do it here because if it fails, the transaction fails rather than being invalid.
+    verifyCertificateSignature(trs)
+    # Process cross-chain messages in inbox update.
+    crossChainMessages = trs.params.inboxUpdate.crossChainMessages
+
+    # First process basic checks for all CCMs.
+    for ccmBytes in crossChainMessages:
+        # Verify general format. Past this point, we can access ccm root properties.
+        try:
+            validateFormat(ccm)
+            # The CCM must come from the sending chain.
+            if trs.params.sendingChainID != ccm.sendingChainID:
+                raise Exception("CCM is not from the sending chain.")
+        except:
+            terminateChain(trs.params.sendingChainID)
+            # In this case, we do not even update the chain account with the new certificate.
+            return
+
+    for ccmBytes in crossChainMessages:
+        try:
+            ccm = decode(crossChainMessageSchema, ccmBytes)
+            # Modules can verify the CCM. In particular, this calls the interoperability verifyCrossChainMessage defined above.
+            # The Token module verifies the escrowed balance in the CCM sending chain.
+            for each module mdl for which verifyCrossChainMessage exists:
+                mdl.verifyCrossChainMessage(trs, ccm)
+        except:
+            # Chain was already terminated.
+            if terminatedStateAccount(chainID) does not exists:
```

Comment:
I think this check is not necessary any more as it is done already in the `terminateChain` function.

Reply:
Also `chainID` is not defined.

Reply:
removed